helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -913,6 +935,76 @@ private RollingUpdateOp rollingUpdateDeploy(final ZooKeeperClient client,
     }
   }
 
+  private RollingUpdateOp rollingUpdateAwaitStopped(final ZooKeeperClient client,
+                                                    final RollingUpdateOpFactory opFactory,
+                                                    final DeploymentGroup deploymentGroup,
+                                                    final String host) {
+    final TaskStatus taskStatus = getTaskStatus(client, host, deploymentGroup.getJobId());
+    final JobId jobId = deploymentGroup.getJobId();
+
+    if (taskStatus == null) {
+      // Agent has not written job status to zookeeper.
+
+      if (getDeployment(host, jobId) == null) {
+        // The job is not listed under /config/hosts. It may have been manually undeployed?
+        // We can't undeploy it if it's not there anymore, so move on.
+        return opFactory.nextTask();
+      }
+
+      // Check if we've exceeded the timeout for the rollout operation.
+      if (isRolloutTimedOut(client, deploymentGroup)) {
+        return opFactory.error("timed out while retrieving job status", host,
+                               RollingUpdateError.TIMED_OUT_RETRIEVING_JOB_STATUS);
+      }
+
+      // We haven't detected any errors, so assume the agent will write the status soon.
+      return opFactory.yield();
+    }
+
+    if (!taskStatus.getState().equals(TaskStatus.State.STOPPED)) {
+      // job isn't stopped yet.
+
+      // We exceeded the configured deploy timeout, and this job is still not stopped
+      if (isRolloutTimedOut(client, deploymentGroup)) {
+        return rollingUpdateTimedoutError(
+            opFactory, host, jobId, taskStatus, TaskStatus.State.STOPPED);
+      }
+
+      return opFactory.yield();
+    }
+
+    return opFactory.nextTask();
+  }
+
+  private RollingUpdateOp rollingUpdateMarkUndeployed(final ZooKeeperClient client,
+                                                      final RollingUpdateOpFactory opFactory,
+                                                      final DeploymentGroup deploymentGroup,
+                                                      final String host) {
+    final List<String> hostsToUndeploy = getHosts(
+        client, Paths.statusDeploymentGroupRemovedHosts(deploymentGroup.getName()));
+
+    if (!hostsToUndeploy.removeAll(ImmutableList.of(host))) {
+      // Something already removed this host.
+      return opFactory.nextTask();
+    }
+
+    try {
+      // TODO(negz): Are we susceptible to a race?
+      // Master A reads list
+      // Master B reads list
+      // Master A removes host X
+      // Master B removes host Y
+      // Master A commits list without X and with Y
+      // Master B commits list without Y and with X
+      // Does failing the transaction using check() fail the whole update, or will we retry this op?
```

Comment:
This should not happen, because two masters cannot work on the same rolling update at the same time, right?